thrust/zip_function.h

```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#include <thrust/type_traits/integer_sequence.h>
+#include <thrust/detail/type_deduction.h>
+
+THRUST_BEGIN_NS
+
+namespace zip_detail {
+
+template <typename Function, typename Tuple, std::size_t... Is>
+__host__ __device__
+decltype(auto) apply_impl(Function&& func, Tuple&& args, index_sequence<Is...>)
+{
+  return func(thrust::get<Is>(THRUST_FWD(args))...);
+}
+
+template <typename Function, typename Tuple>
+__host__ __device__
+decltype(auto) apply(Function&& func, Tuple&& args)
+{
+  constexpr auto tuple_size = thrust::tuple_size<typename std::decay<Tuple>::type>::value;
+  return apply_impl(THRUST_FWD(func), THRUST_FWD(args), make_index_sequence<tuple_size>{});
+}
+}
+
+// Adaptor type that turns an N-ary function object into one that takes a tuple
+// of size N so it can easily be used with algorithms taking zip iterators
+template <typename Function>
+class zip_function
+{
+  public:
+     __host__ __device__
+    zip_function(Function func) : func(std::move(func)) {}
```

Comment:
I think `zip_function` may need a copy constructor.

Reply:
It should be implicitly declared since the `zip_function(Function)` constructor isnt a catch-all template